torchax/torchax/view.py

```diff
@@ -0,0 +1,384 @@
+
+    def update_tensor(self, new_value: jax.Array, jax_array: jax.Array) -> jax.Array:
+        raise NotImplementedError("AsStridedInfo.update not implemented")
+
+    def calculate_output_shape(self, source: jax.Array) -> List[int]:
+        raise NotImplementedError(
+            "AsStridedInfo.calculate_output_shape not implemented"
+        )
+
+
+class DiagonalInfo(ViewInfo):
+    """
+    Information for diagonal operations.
+    Extracts diagonal elements from a tensor.
+    """
+
+    def __init__(self, offset: int = 0, dim1: int = 0, dim2: int = 1) -> None:
+        """
+        Args:
+            offset: Offset from the main diagonal
+            dim1: First dimension for diagonal extraction
+            dim2: Second dimension for diagonal extraction
+        """
+        super().__init__(ViewInfoType.DIAGONAL)
+        self.offset: int = offset
+        self.dim1: int = dim1
+        self.dim2: int = dim2
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, DiagonalInfo):
+            return False
+        return (
+            self.offset == other.offset
+            and self.dim1 == other.dim1
+            and self.dim2 == other.dim2
+        )
+
+    def apply(self, jax_array: jax.Array) -> jax.Array:
+        raise NotImplementedError("DiagonalInfo.apply not implemented")
+
+    def update(self, new_value: jax.Array, parent: jax.Array) -> jax.Array:
+        raise NotImplementedError("DiagonalInfo.update not implemented")
+
+    def calculate_output_shape(self, source: jax.Array) -> List[int]:
+        raise NotImplementedError("DiagonalInfo.calculate_output_shape not implemented")
+
+
+class View(torch.Tensor):
+    """
+    A View is a reference to another Tensor or another View,
+    with a transformation applied to it.
+    """
+
+    @staticmethod
+    def __new__(
+        cls, parent: Union["torchax.Tensor", "View"], view_info: ViewInfo, env: Any
+    ) -> "View":
+        """
+        Args:
+            parent: Parent tensor or view
+            view_info: Information about the view transformation
+            env: Environment for tensor operations
+        """
+        shape = view_info.calculate_output_shape(parent.jax())
+        return torch.Tensor._make_wrapper_subclass(
+            cls,
+            shape,
+            device="meta",
+            dtype=parent.dtype,
+            requires_grad=False,
+        )
+
+    def __init__(
+        self, parent: Union["torchax.Tensor", "View"], view_info: ViewInfo, env: Any
+    ) -> None:
+        super().__init__()
+        self.parent = parent
+        self.view_info = view_info
+        self._env = env
+
+    def get_transformation_chain(self) -> List[ViewInfo]:
+        """
+        Get all view transformations from the source tensor to this view.
+        """
+        if isinstance(self.parent, View):
+            return self.parent.get_transformation_chain() + [self.view_info]
```

Comment:
this impl is quadratic on number of ancestors.
You can assign the result to a variable and just use `.append` instead.